### ISSUE TITLE
Implement Debugging/Editor Interface

### ISSUE THREAD

**Issue by controllerface** [controllerface/bvge#11]
Now that the basic CRUD API is working, more complex systems can start to be developed around it. However, the need for a solid debugging and/or "level editing" UI will rapidly become important.

After doing several tutorials, and just my own experience with level editing UIs in the past, I have decided I don't want to implement the traditional level editor view and I would also like to avoid IMGui if possible as it adds overhead both in that I would need to get more familiar with its paradigms, and also I don't really want more dependencies.

Instead, I plan to leverage my existing skill with server/client dev, and the other experiments I have done in this area (like Commander J and Autohammer) to build a browser based debugging endpoint. The general idea is that the game is started up with a flag that enables a web endpoint on localhost, which when connected to in a browser, will display debugging information.

In this first iteration, the only need is for simple debugging information. It really only needs to be PoC level, so I am not even that concerned with aesthetics, a simple CSS Grid with some real-time data is all that is needed. Because this endpoint is **_never_** intended to be exposed outside of localhost, there will be no need for TLS connectivity, and in fact it should be avoided due to the added latency. 

I have not yet decided if I want to use websockets for bi-directional communication, or go with the route I have gone before, which is standard REST for incoming communication, and an eventsource connection for outbound events. Because I envision this as being very much an event driven system, I will probably go with the later, but will still give websockets a look.

It will be vital that data is sent out asynchronously to the UI and the impact on the engine itself must be kept to an absolute minimum. If possible, I want to have an internal debugger thread (only active when the debug mode flag is enable) that handles getting data from the engine in a way that doesn't interfere with the game loop. It is OK if data lags behind the renderer by a small amount, but I am hopeful I can keep the lag under a second, or under 500ms ideally.

The client side portion will **not use any frameworks** and this is a hard requirement. All UI will be written using standard JS and Web Components and normal CSS. I will probably also hand-write the sever side portion, I am on the fence if I should use Jetty again. While it does work well, it's also significantly more featured than this endpoint will need.

**Comment by controllerface** [comment]
thinking about this one, I think I may wait a bit before tackling this. I started working on it and implementing the debugger as a GameSystem, but had second thoughts and looked at moving it higher up, but also I realize there's a quite a few things that I haven't done yet and I am not sure how they will work so I don't want to paint myself into a corner. Will put this back into the backlog for now.

**Comment by controllerface** [comment]
I have decided that for a change of pace, i will start poking at this again. Also, I realize that having some way to debug data in real-time would really be helpful for many tasks I have ahead of me. 

May plan is to implement a simple plaintext http and websocket server that can be used as a local debugging interface. Once it is in place, I will start leveraging the work from Commander J to build a basic GUI.

**Comment by controllerface** [comment]
OK, so I misstated my intent to make a WebSocket server, what i meant was an EventSource server 😄 Which is now working.

At the moment, there isn't any UI at all, just a basic html/js/css combo, with the js file simply opening an event source connection and doing nothing else. I will probably not go crazy implementing a full fledge UI with all kinds of crazy effects, but I would like at least a simple control-panel of sorts, that can show me whatever data has been sent tot he UI.

For now, I have implemented this so it is on a 1 second interval, and it only sends whatever the current data value is for the event when that second has elapsed. I know from previous experience that sending data too quickly just makes the UI go nuts on repaint if there's lots of elements. What I will likely do is make some kind of "stat entry" element, and whenever a new event comes in, if it has a matching element already, I will update it, if not i will create it with the new value. 

Once I have something like that up and running I will probably call this "done" even though it likely will get some fleshing out further down the line. I have named it "Editor" rather that "Debugger" as i do expect i will find some use for it for editing levels and possibly assets, characters, or objects. But for now, it really is going to be just for debugging.

I may implement some type of "immediate message" method that allows some events to bypass the 1 second interval as well, as that probably will be helpful for some stuff that happens really quickly.
